rust/src/sans_io/read.rs

```diff
@@ -0,0 +1,1014 @@
+                    };
+                }
+                // A chunk record can have more bytes after the `data` member, which we need
+                // to discard.
+                PaddingAfterChunk { len } => match self.consume(len)? {
+                    Span(_) => {
+                        if let Some(hasher) = &mut self.data_section_hasher {
+                            self.calculated_data_section_crc = Some(hasher.clone().finalize());
+                        }
+                        self.currently_reading = Record;
+                    }
+                    Remainder(need) => return self.request(need),
+                },
+            }
+        }
+    }
+
+    // load `amount` bytes into the uncompressed data buffer, returning the remainder if more
+    // needs to be loaded.
+    fn load(&mut self, amount: usize) -> McapResult<SpanOrRemainder> {
+        let want = match self.uncompressed.span(amount) {
+            Span(b) => return Ok(Span(b)),
+            Remainder(want) => want,
+        };
+        match &mut self.from {
+            File => Ok(Remainder(want)),
+            Chunk(chunk_state) => {
+                let decompressor = match &mut chunk_state.decompressor {
+                    None => return Ok(Remainder(want)),
+                    Some(dec) => dec,
+                };
+                self.uncompressed.data.resize(
+                    std::cmp::max(self.uncompressed.data.len(), self.uncompressed.end + want),
+                    0,
+                );
+                let src = self.compressed.unread();
+                if src.is_empty() {
+                    return Ok(Remainder(std::cmp::min(
+                        clamp_to_usize(chunk_state.compressed_remaining),
+                        DEFAULT_CHUNK_DATA_READ_SIZE,
+                    )));
+                }
+                let dst = self.uncompressed.tail();
+                let res = decompressor.decompress(src, dst)?;
+                let newly_decompressed = &self.uncompressed.tail()[..res.wrote];
+                if let Some(hasher) = &mut chunk_state.hasher {
+                    hasher.update(newly_decompressed);
+                }
+                self.uncompressed.mark_written(res.wrote);
+                self.compressed.mark_read(res.consumed);
+                chunk_state.compressed_remaining -= res.consumed as u64;
+                if self.compressed.len() == 0 {
+                    self.compressed.reset();
+                }
+                match self.uncompressed.span(amount) {
+                    Span(b) => Ok(Span(b)),
+                    Remainder(_) => Ok(Remainder(std::cmp::min(
+                        clamp_to_usize(chunk_state.compressed_remaining),
+                        res.need,
+                    ))),
+                }
+            }
+        }
+    }
+
+    // Consume `amount` bytes of the uncompressed input buffer if enough is available. On failure,
+    // return the extra amount required as an error value.
```

Comment:
why does consume not :just: produce a ReadAction when it needs more input?

Reply:
The answer is the borrow checker fights real hard when you do that. I need to come up with the mutable return buffer right at the moment of returning, otherwise too many mutable references are in scope at the same time.